`core::marker::NoCell` in bounds (previously known an `Freeze`) (#3633)
rust-lang/rfcs: text/3633-freeze-in-bounds.md

Note that for this latter use-case, `core::marker::Freeze` isn't entirely sufficient, as an additional proof that `T` doesn't contain padding bytes is necessary to allow this transmutation to be safe, as reading one of `T`'s padding bytes as a `u8` would be UB.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

`core::marker::NoCell` is a trait that is implemented for any type whose memory layout doesn't contain any `UnsafeCell`: it indicates that the memory referenced by `&T` is guaranteed not to change while the reference is live.

It is automatically implemented by the compiler for any type that doesn't contain an un-indirected `core::cell::UnsafeCell`.

Notably, a `const` can only store a reference to a value of type `T` if `T: core::marker::NoCell`, in a pattern named "static-promotion".

> As `core::marker::NoCell` is an auto-trait, it poses an inherent semver-hazard (which is already exposed through static-promotion): this RFC proposes the simultaneous addition and stabilization of a `core::marker::PhantomCell` type to provide a stable mean for maintainers to reliably opt out of `NoCell` without forbidding zero-sized types that are currently `!NoCell` due to the conservativeness of `NoCell`'s implementation being locked into remaining `!NoCell`.

Reviewer: This paragraph is one sentence, would it be easier to understand this way?

```suggestion
As `core::marker::NoCell` is an auto-trait, it poses an inherent semver-hazard (which is already exposed through static-promotion). This RFC proposes the simultaneous addition and stabilization of a `core::marker::PhantomCell` type, to provide a stable means for maintainers to reliably opt out of `NoCell`, without forbidding zero-sized types. These types are currently `!NoCell` due to the conservativeness of `NoCell`'s implementation.
```

This means that their byte representation cannot change as long as a reference to them exists.

Note that `T: NoCell` is a shallow property: `T` is still allowed to contain interior mutability,
provided that it is behind an indirection (such as `Box<UnsafeCell<U>>`).
Notable `!NoCell` types are [`UnsafeCell`](core::cell::UnsafeCell) and its safe wrappers
such as the types in the [`cell` module](core::cell), [`Mutex`](std::sync::Mutex), and [atomics](core::sync::atomic).
Any type which contains a non-`NoCell` type without indirection also does not implement `NoCell`.

`T: NoCell` is notably a requirement for static promotion (`const REF: &'a T;`) to be legal.

Note that static promotion doesn't guarantee a single address: if `REF` is assigned to multiple variables,
they may still refer to distinct addresses.

Whether or not `T` implements `NoCell` may also affect whether `static STATIC: T` is placed
in read-only static memory or writeable static memory, or the optimizations that may be performed
in code that holds an immutable reference to `T`.

# Semver hazard
`NoCell` being an auto-trait that encodes a low level property of the types it is implemented for,

> it is strongly advised against to rely on external types maintaining that property unless that

Reviewer: ```suggestion
you should avoid relying on external types maintaining that property, unless that
```

contract is explicitly stated out-of-band (through documentation, for example).

Conversely, authors that consider `Freeze` to be part of a type's contract should document this
fact explicitly.

## The ZST caveat
While `UnsafeCell<T>` is currently `!Freeze` regardless of `T`, allowing `UnsafeCell<T>: Freeze` if `T` is
a Zero-Sized-Type is currently under consideration.

Therefore, the advised way to make your types `!Freeze` regardless of their actual contents is to add a
[`PhantomNotFreeze`](core::marker::PhantomNotFreeze) field to it.

[`PhantomData<T>`](core::marker::PhantomData) only implements `Freeze` if `T` does, making it a good way

> to conditionally remove a generic type's `Freeze` auto-impl.

Reviewer 2: The change this comment is about is phrased as "currently under consideration", so I assume that it is not meant as part of this RFC(?).

-----

I think that having `UnsafeCell<Zst>` and/or `[NotFreeze; 0]` (which is mentioned elsewhere but not here) be `Freeze` is inconsistent and not useful, especially in combination with `PhantomData<NotFreeze>` being `!Freeze`.

Saying "`UnsafeCell<T>` is never `Freeze`" is quite simple. Saying "`UnsafeCell<T>` is never `Freeze`, unless `T` is zero-sized" is less simple, harder to check for generically (currently), and I don't see a strong motivation for it.

Also, how would "`PhantomData<T>` inherits `Freeze`ness"/`PhantomNotFreeze` and "`UnsafeCell<Zst>: Freeze`" interact? It seems like it would be very confusing if `UnsafeCell<PhantomNotFreeze>: Freeze`, but to make that not the case, the rule would have to be something like "`UnsafeCell<T>` is never `Freeze`; unless `T` is zero-sized, then it is `Freeze`; unless `T` is `PhantomNotFreeze` or `PhantomData<U>` where `U` is not `Freeze`, then it is not `Freeze` again".

"`UnsafeCell<Zst>: Freeze`" could also be a semver/cargo-feature hazard, as it is possible to have a type be zero-sized under some cargo features and not others (e.g. error types that keep additional state under debugging features, like [`std::cell::BorrowError`](https://doc.rust-lang.org/nightly/src/core/cell.rs.html#744)), so `UnsafeCell<ThatType>` would be sometimes `Freeze` and sometimes not.

For `[T; 0]`, I think this would interact negatively with future resolutions to ["the `&Header` problem"](https://github.com/rust-lang/unsafe-code-guidelines/issues/256)[^1]. Concretely, I think it would be confusing if `#[repr(C)] struct ThinSlice<T> { length: usize, tail: [T; 0] }` was `Freeze` even if `T` wasn't.

Also: I mentioned in another thread that *removing* `Freeze` from `PhantomData` shouldn't cause any miscompilations. Conversely, *adding* `Freeze` to `UnsafeCell<Zst>` or `[NotFreeze; 0]` could introduce miscompilations[^1]. We could say that "any code that was relying on `UnsafeCell<Zst>: !Freeze` had UB", but that seems a much higher bar of breaking change than the `PhantomData` change.

[^1]: Assuming the compiler continues to use `Freeze` for its own aliasing reasoning and not something stricter.

Author: Terribly interesting comment, and I think I agree on basically all points. 

This RFC indeed doesn't attempt to take any position on that debate, simply warning in docs that the debate does exist; in the sole goal of earning users that want to opt out of `Freeze` that some techniques that would work now to do so may not work forever.

Reviewer 3: > We could say that "any code that was relying on UnsafeCell<Zst>: !Freeze had UB", but that seems a much higher bar of breaking change than the PhantomData change.

I think we have been very clear that relying on `UnsafeCell` to do anything outside of the contents of that type is not permitted.

# Safety
This trait is a core part of the language, it is just expressed as a trait in libcore for
convenience. Do *not* implement it for other types.
```

Mention could be added to `UnsafeCell` and atomics that adding one to a previously `NoCell` type without an indirection (such as a `Box`) is a SemVer hazard, as it will revoke its implementation of `NoCell`.

## Fixing `core::marker::PhantomData`'s `NoCell` impl

At time of writing, `core::marker::PhantomData<T>` implements `NoCell` regardless of whether or not `T` does.

This is now considered a bug, with the corrected behaviour being that `core::marker::PhantomData<T>` only implements `NoCell` if `T` does.

> While crates that would "observe" this change exist, the current consensus is that it would only break invalid usages of that invariable bound.

Reviewer 4: Could someone provide some more context for why this is considered a "bug"? My only interaction with this trait is through the lense of Project Safe Transmute, and I very well might be unaware of other use-cases where this might be "buggy". But, skimming the Motivation section, I didn't see a motivation for this change.

For transmutation, the valuable property is that *a type's immediate bytes cannot be mutated through a shared reference*. `PhantomData<T>` clears this bar as does [`Exclusive<T>`](https://doc.rust-lang.org/stable/std/sync/struct.Exclusive.html). Zerocopy adopts these semantics in its [`Immutable`](https://docs.rs/zerocopy/latest/zerocopy/trait.Immutable.html) trait, and — at the time of writing — [`mem::TransmuteFrom`](https://doc.rust-lang.org/stable/std/mem/trait.TransmuteFrom.html) internally leverages `Freeze` to enforce these semantics, too.

The Motivation section says the "`zerocopy` intends to replace its bespoke trait with a re-export of `core::marker::Freeze`". This was true, but I'm not sure this new `NoCell` is quite as expressive as we'd like. On the compiler safe transmute side of things, I'd prefer if the semantics of `Freeze` were not modified (since we rely on them), but rather that proposed `NoCell` trait was added in parallel.

Reviewer 5: Yeah, sounds like another reason for opt-in trait. The semantics of current `!Freeze` is "this type actually physically contains at least one byte in `UnsafeCell`" and is used by the compiler to perform optimizations wile the new semantics would be "`unsafe` code can not rely on this type not containing `UnsafeCell`". As a result, people who defensively use `PhantomCell` just to have API flexibility would also drop those optimizations jut because they want to be extra careful.

Reviewer 3: From an opsem perspective, there are cases where need to know whether bytes that are *outside* of the type's size should be considered interior mutable nor not (this occurs in the context of https://github.com/rust-lang/unsafe-code-guidelines/issues/256). I think `PhantomData<Cell<T>>` or `UnsafeCell<()>` are a good way to indicate "this type points to interior mutable data but it's not in any of the fields you can see".

I don't know if that concern played any rule in the discussion though, it is pretty niche. I do find it questionable to consider a type to be "not interior mutable" when the programmer clearly indicated that "this acts as-if there's a `T` here". If the programmer meant to say "this acts as-if we own a `T` but it's stored out-of-line", they should have written `PhantomData<Box<T>>`, which is always `Freeze`/`NoCell`.

Reviewer 3: > The semantics of current !Freeze is "this type actually physically contains at least one byte in UnsafeCell" and is used by the compiler to perform optimizations 

It is used for more than that, as the RFC explains. If the trait was only relevant for optimizations, @p-avital wouldn't need this RFC to make their crate work the way it did.

Reviewer 4: > From an opsem perspective, there are cases where need to know whether bytes that are _outside_ of the type's size should be considered interior mutable nor not (this occurs in the context of [rust-lang/unsafe-code-guidelines#256](https://github.com/rust-lang/unsafe-code-guidelines/issues/256)).

That's extremely neat.

> But I do find it questionable to consider a type to be "not interior mutable" when the programmer clearly indicated that "this acts as-if there's a T here".

Good point. That's sensible UX for an auto trait. For zerocopy, that hasn't been an issue because we require `#[derive(Immutable)]`.

> How does safe transmute rely on the semantics of Freeze/NoCell...

For `mem::TransmuteFrom`, we emit `Freeze` bounds for `T` and `U` anytime we have to query for a transmutation from a `&T` to an `&U`. And since transmutability is evaluated recursively, the surface-level experience of this is something like "a recursive-through-reference-indirection Freeze".

> and what would be the downside of the proposed change in the context of safe transmute?

For example `&PhantomData<T>: TransmuteFrom<&PhantomData<U>>` presently holds; this would no longer hold if we used `NoCell` instead of `Freeze`.

For zerocopy, we'd like to implement `Immutable` for `Exclusive`, once `Exclusive` is stabilized. I haven't seen discussion about `Exclusive` here, but the name `NoCell` sorta implies that it *wouldn't* be implemented for `Exclusive<Cell<T>>`.

Reviewer 3: I also find it quite concerning that anyone would rely on the corner cases of `Freeze` that much. The trait's existence is entirely incidental, it was never subject to proper design; we used to have code recursively traversing a type to ind `UnsafeCell` and the trait is just a neater way of doing the same thing. Do not use its current behavior as yardstick for anything relevant; instead, consider from first principles what you actually want.

Reviewer 4: > it seems entirely reasonable it would also be sensitive to `Freeze`/`NoCell` (including inside `PhantomData`).

The design goal is that for any two types `Src`, `Dst`, `Dst: TransmuteFrom<Src>` if(f) it is sound to reinterpret the bytes of all possible values of `Src` as a `Dst`. We care about corner cases like `&PhantomData<T>: TransmuteFrom<&PhantomData<U>>`, because each additional corner case where `TransmuteFrom` falls short of that goal gives folks another reason to believe that *they know better than* `TransmuteFrom`.

> I also find it quite concerning that anyone would rely on the corner cases of Freeze that much.

Why? `TransmuteFrom` is an experimental, very-much-in-development lang item. The compiler *is* `Freeze`'s only customer. Right now, it does what we need it to. If its semantics change and this is no longer true, we'll use something else.

> I don't know what Exclusive is.

https://doc.rust-lang.org/stable/std/sync/struct.Exclusive.html

Reviewer 3: > The design goal is that for any two types Src, Dst, Dst: TransmuteFrom<Src> if(f) it is sound to reinterpret the bytes of all possible values of Src as a Dst

The `PhantomData` might be meaningful and relevant for soundness, so you can't just transmute it away.

> We care about corner cases like &PhantomData<T>: TransmuteFrom<&PhantomData<U>>, because each additional corner case where TransmuteFrom falls short of that goal gives folks another reason to believe that they know better than TransmuteFrom.

By the same token you should allow transmuting `&UnsafeCell<()>` to `&PhantomData<T>` and to `&[u8; 0]`, which sounds like a bad idea to me. I'm sure I can come up with a long list of legal transmutes that are correct for more and more intricate reasons, and it would be a fool's errand to try and get the compiler to understand them all. The exact shape of the set of legal transmutes is extremely intricate and it's not worth trying to exactly redraw that line with `TransmuteFrom` -- it is entirely expected that a type system "cuts off" sound programs for the sake of making the system easier to reason about or easier to implement (or even possible to implement as the original question is likely undecidable anyway).

> https://doc.rust-lang.org/stable/std/sync/struct.Exclusive.html

I see, thanks. If making `Exclusive: NoCell` sound, we could add an `unsafe impl`. It is far from clear to me that that is sound though.

---

Taking a step back: the entire point of the proposed changes around `PhantomData` is to make the trait more suitable for use in a public API. Using `Freeze` in its current form for safe-transmute has all the same downsides that the RFC specifically attempts to address.

Anyway, if the `NoCell` in this RFC (which is primarily about const promotion) does not suit safe-transmute, a new trait can be added if it is sufficiently well-motivated. "We happen to like the shape of that internal thing that accidentally existed" is not sufficient motivation for using a semver-unstable implementation-detail-leaking trait in new APIs.

Reviewer 4: <details>
<summary>safe transmute digression</summary>

> The `PhantomData` might be meaningful and relevant for soundness, so you can't just transmute it away.

I wasn't giving you an example of a `Src` and `Dst` that contain a `PhantomData`. I *literally* mean `&PhantomData<Src>` and `&PhantomData<Dst>`. There is no transmute between these two values that would be unsound.

> it would be a fool's errand to try and get the compiler to understand them all

That's the errand we're on. It hasn't been too difficult. :-)

</details>

---

> I see, thanks. If making `Exclusive: NoCell` sound, we could add an `unsafe impl`. It is far from clear to me that that is sound though.

This seems like a question that needs to be resolved, then. The safety condition of this trait — let's call it `BikeshedIcey` — should be crystal clear about when a type satisfies its condition (even if compiler limitations prevent us from providing an overlapping impl).

If `BikeshedIcey` adopts the safety condition that [`zerocopy::Immutable`](https://docs.rs/zerocopy/latest/zerocopy/trait.Immutable.html) has:

> T: Immutable indicates that T does not permit interior mutation, except by ownership or an exclusive (&mut) borrow.

...then it's fairly clear that [`Exclusive<T>`](https://doc.rust-lang.org/stable/std/sync/struct.Exclusive.html) and [`PhantomData<T>`](https://doc.rust-lang.org/stable/std/marker/struct.PhantomData.html) meet that bar, as well as types like `[T; 0]` (even though we can't provide that `impl` [until `#[marker]` is stabilized](https://github.com/rust-lang/rust/issues/29864), but we do intend to do so when possible).

If, alternatively, `BikeshedIcey` adopts a safety condition that requires `PhantomData<T>` to be conditionally `BikeshedIcey`, then we need to come up with a clear safety condition that's just as clear. It should be obvious why `PhantomData<T>` must be conditionally `BikeshedIcey`, without it being shocking that `&T` is *un*conditionally `BikeshedIcey`. It should be obvious whether or not [`Exclusive<T>`](https://doc.rust-lang.org/stable/std/sync/struct.Exclusive.html) meets this condition.

So: what's the safety invariant we're trying to provide with this trait? (I'm not posing this question literally to you, @RalfJung, but more broadly to @p-avital and the lang team.)

Reviewer 3: Note that the behavior of `PhantomData`  does not follow from the safety invariant. For instance, nothing about the invariant of `Send` or `Sync` imples anything about `PhantomData`. Rust would be entirely sound if, when `PhantomData` was added, we had decided that it is always `Send` and `Sync` (but some things that are possible to day might be impossible to express in this alternative universe). `PhantomData` is a purely *syntactic* marker used to control auto trait inference (and dropck); semantic questions like the exact invariant of some trait do nit have any direct bearing on this type.

`PhantomData` as a type itself isn't even very relevant, which is why I pushed back against your using it as a bare type above. Its role is entirely as a *field* in some other type that involves unsafe code. It's basically an annotation, disguised as a type that we put on a field. The meaning of the annotation `PhantomData<T>` is to make the outer type *conditionally* Send/Sync/... only if `T` has that trait. In that sense it's just syntactic sugar for writing an impl like that directly.

Hence, I see no reason why `NoCell` would behave any different from all the other auto traits here.

---

Regarding the safety invariant of that trait, its main purpose in this RFC is to control whether const promotion can occur. So that boils down to, "it must be okay to store a value of this type in immutable memory and hand a shared reference to that memory out to users". All ZST trivially satisfy this property. But again, that doesn't help to determine what `PhantomData` should do.

Reviewer 4: > its main purpose in this RFC is to control whether const promotion can occur

That's a valid use-case, but the RFC spends a lot of prose on the transmutation uses of this trait. I'm wondering if perhaps we've been trying to cram too many use-cases into a single trait. The affordances that are useful for const promotion are different than the affordances that are useful for transmutation.

@traviscross, @p-avital, if we restrict the use-case of this trait to just controlling const promotion, the naming discussions get a *lot* simpler — e.g., `ConstPromotable` is direct and not misleading. (`NoCell`, by contrast, is a compromise when it comes to cases like `PhantomData<T>` and `[T; 0]`, and doesn't lend itself to an intuition about whether it can be implemented for `Exclusive<T>`.)

Project Safe Transmute can then, as @RalfJung suggests, introduce a separate trait that captures the immutability invariant we need.

Reviewer 3: > the RFC spends a lot of prose on the transmutation uses of this trait

Ah fair, I've not looked at the text in a bit and forgot about this.

> if we restrict the use-case of this trait to just controlling const promotion, the naming discussions get a lot simpler

Hm... if we want to pick a different name for that reason now, that basically means we have to explore whether we really want a separate meaning for transmutation-related concerns. I think the answer to that is "no", but I was hoping we could avoid having to resolve this now. Oh well, I guess we cannot avoid it.

Reviewer 4: We have to figure out what purpose this trait serves and what its safety condition is *now* (prior to stabilization), regardless of what we name it. In the interest of haste, I'd suggest *not* blocking stabilization of this trait on unifying two different purposes with two slightly different needs. Controlling const promotion (and preventing const promotion regressions) a compelling reason, on its own, for this trait to be stabilized.

But do consider giving the trait a name that reflects this purpose, and won't cause a headache later on with safe transmute. If the semantics of the two traits ends up being the same, then we can just alias `trait Immutable = ConstPromotable`, or whatever.

Reviewer 4: No strong opinion, but I would like to offer this caveat for when you discuss this with the lang team: I think `*Promote` risks giving the impression that types implementing that trait *will* be const promoted, whereas `*Promotable` conveys that implementing types *may* be const promoted. Verb names are great when it's the user doing a verb, but I'm not sure they're as appropriate when it's the compiler *maybe* doing the verb.

Reviewer 3: > If the semantics of the two traits ends up being the same, then we can just alias trait Immutable = ConstPromotable, or whatever.

That means we'd be blocked on trait aliases.
I opened a [Zulip thread](https://rust-lang.zulipchat.com/#narrow/channel/213817-t-lang/topic/The.20role.20of.20Freeze.20in.20TransmuteFrom/with/508570722) to hopefully dig a bit deeper into why safe-transmute would want a version of `NoCell` where `PhantomData` is always `NoCell`.

Reviewer 4: @RalfJung No, we're not blocked on trait aliases in any meaningful sense. For `mem::TransmuteFrom`, `Freeze` is an implementation detail — not part of the public API. Zerocopy can continue to use `#[derive(Immutable)]` indefinitely with no significant disadvantages, and some advantages, for example:
- the SemVer stability of derivable traits is better understood by the community
- we can tailor the documentation to transmutability, not const promotability
- we can make implementation decisions that suit transmutability, independent of const promotability; e.g., implementing `Immutable` unconditionally for [`Exclusive<T>`](https://doc.rust-lang.org/stable/std/sync/struct.Exclusive.html), or for `[T; 0]`

I don't think we're ready to commit to saying that the affordances that are useful for transmutability are the same that are useful for const promotability, and I don't think we should block this RFC on that commitment. Even if it *did* turn out to be the case that we desperately wanted to write `trait Immutable = ConstPromotable` after stabilization, we can do so without trait aliases by writing `pub use ConstPromotable as Immutable`

Reviewer 6: +1 to narrowing the scope to const promotability. I agree with @jswrenn that we aren't yet ready to commit to particular semantics for whatever trait we end up using in `mem::TransmuteFrom`. Including `mem::TransmuteFrom` in the scope of this RFC will result in one of two outcomes:
- We risk committing to something that we later learn has an impedance mismatch with what `mem::TransmuteFrom` actually needs
- We stall this RFC for a long time, roughly until `mem::TransmuteFrom` is ready for stabilization, which will likely be 2026 at the earliest

Reviewer 7: Narrowing the scope to allow for more flexibility sounds nice, but we are talking about stabilizing a new auto trait, with all the hazards it implies of implicit propagation and semver. Auto traits are, at least for now, an implicit surface area of every public struct and and RPIT usage. So we should strongly favor a solution that introduces one rather than two.

Reviewer 7: > To get back to verbs, how about StaticPromote/ConstPromote?

Your use case right now relies on promotion, but isn't it true that if we were designing a first-class generic statics feature, we would want to use the same trait on generic parameter bounds? So if we were going to pick a name for this use case only, it should be something that makes sense for generic statics too.

One thought that occurs to me is that statics get "baked"... or, ahem, even "frozen"... into a binary.

Reviewer 4: > So we should strongly favor a solution that introduces one rather than two.

I don't think we need an additional trait anytime soon (if ever). Zerocopy gets real value out of `derive(Immutable)`. `mem::TransmuteFrom` doesn't use `Freeze` in its public API (and can switch to some other internal mechanism), and is documented to *not* be SemVer stable (it's a building block for SemVer stable APIs, but it is not itself SemVer stable).

Reviewer 4: > So we should strongly favor a solution that introduces one rather than two.

I don't think we need two traits? Zerocopy [gets a lot of value](https://github.com/rust-lang/rfcs/pull/3633#discussion_r2017175988) out of `#[derive(Immutable)]`, and `mem::TransmuteFrom` only relies on `Freeze` and an implementation detail. I don't see why we would necessarily need to expose that implementation detail, and because `mem::TransmuteFrom` is expressly documented to *not* adhere to SemVer, we expressly would not expect that implementation detail to adhere to SemVer either (if we ever did expose it).

Reviewer 3: If there's only one trait, `NoCell` seems perfectly adequate to me. `PhantomData` can add a "phantom cell", which is less surprising than its current behavior IMO -- it can already add "phantom covariance" and "phantom `!Send`".

Reviewer 8: I looked through Github to see how people use `PhantomData`, and I found that  `PhantomData` is used in three major ways
* to create type-tagged values
    * for example, typed indices into some structure
    * this one is not likely to be relevant to this conversation, since these tags are pretty much always `NoCell`
* to create custom pointers
    * [`Unique`](https://github.com/rust-lang/rust/blob/19f42cb9bb65555e7d95dab7d17573ae6c5d634c/library/core/src/ptr/unique.rs#L37-L45) in `std`
    * [erasable::Thin](https://docs.rs/erasable/latest/erasable/struct.Thin.html)
    * etc. (there are many examples of this)
* to opt-out of auto-traits for forwards compatibility
    * for this use-case it makes more sense to just not implement `NoCell` altogether

When using `PhantomData` with pointers, the data *has* an indirection, so pretty much in all cases you would like the custom pointer to be unconditionally `NoCell`. Under the proposed rules, `Unique` would be conditionally `NoCell`, and this would mean that `Box` is conditionally `NoCell`, which is nonsense. And I would not want some carve-out for `std` to make this specifically work since it affects *all* custom pointers (`Vec`, `Thin`, `Arc`, `triomphe::Rc`, etc.). 

The RFC mentions that you could replace most uses of `PhantomData<T>` with `PhantomData<Ptr<T>>`, but this doesn't work for `Unique` or if you are in a no-alloc context in which case you can't conveniently use `PhantomData<Box<T>>`, which means you would need to use `PhantomData<*const T>` manually implement `Send`/`Sync` if you wanted to keep `NoCell`. This seems like a strict downside for the second most common use-case for `PhantomData`.

I see two ways we could square this:
* Add some way to control the `PhantomData` `NoCell` behavior, i.e. some attribute which opts-in to either behavior (I don't care which one is the default)
* Add a new `PhantomData`-like type which is (un?)conditionally `NoCell`
    * We are already adding more flavors of `PhantomData` for things like `Unpin`,
      and variance. So this doesn't seem unusual for me.

As a final remark, I've only seen one use-case where conditional `NoCell` would be helpful, and only if this [issue](https://github.com/rust-lang/unsafe-code-guidelines/issues/256) is resolved:

Since if you are allowed to access the slice after the bytes of the `Example` when you have `x: &Example`, then `Example` should be conditionally `NoCell`. If you are not allowed to access the slice when given `x: &Example`, then `Example` could be unconditionally `NoCell`.

```rust
#[repr(C)]
struct Example<T> {
    length: usize,
    marker: PhantomData<T>,
    // actually a list of Ts stored after tail, and we guarantee that they are at most aligned to 8-bytes
    tail: [u64; 0],
}
```

I think catering to this use-case isn't worthwhile, since it's extremely niche, so making it a bit more tedious is fine. Aside from something like this, I don't see where conditional `NoCell` would every be helpful.

Reviewer 5: Yeah, that use case is much better modeled with `extern type` rather than tail. And IIRC extern types don't impl auto traits.

Reviewer 6: I've started an [IRLO thread](https://internals.rust-lang.org/t/discussion-distinct-phantom-types-for-different-traits/22654) to discuss the idea of having distinct `Phantom` types so that we can avoid polluting this thread with that discussion.

Reviewer 3: > Under the proposed rules, Unique would be conditionally NoCell

We can just add an `unsafe impl NoCell` to that.
And then the rest of the ecosystem can use `PhantomData<Box<T>>` to represent an owned value behind an indirection. But yeah that won't work in `no_std` context...

> This seems like a strict downside for the second most common use-case for PhantomData.

The flipside is that `NoCell` barely matters for type like this. It controls whether `&const { MyPtrType }` gets promoted, but that's not something people will commonly write. It may control the aliasing model, but again that's only relevant for `&MtPtrType`/`&mut MyPtrType` being passed around, which would be a double-indirection so will also be rather rare.

Reviewer 3: Thinking about this some more....

We could have a type `PhantomDataIndirect` which indicates that this data is logically fully owned by this type, but stored "somewhere else", not inline in this type. Since `NoCell` is about the immediate data stored in this type, that would explain why `PhantomDataIndirect` should be unconditionally `NoCell`. That type would probably be an easier sell than an entire new pointer type with all its associated API surface.

I would frame your proposal as saying that `PhantomData` *should* mean `PhantomDataIndirect`, since that's a common use-case (for custom pointers).

To implement `PhantomDataIndirect` on top of `PhantomData` or vice versa, we need `unsafe impl NoCell`, which user code cannot do. So maybe we just need both of these types?

Reviewer 9: We could add a `struct ForceCell<T: ?Sized>(!)` that unconditionally implements `NoCell` (and conditionally impls all other auto traits) but that can't actually be constructed.  Then it could only be used with `PhantomData`.  That would seem to have many fewer questions than adding a new full-blown pointer type.

Reviewer 9: We could add a `struct ForceNoCell<T: ?Sized>(!)` that unconditionally implements `NoCell` (and conditionally impls all other auto traits) but that can't actually be constructed.  Then it could only be used with `PhantomData`.  That would seem to have many fewer questions than adding a new full-blown pointer type.

Reviewer 3: So then, `PhantomDataIndirect<T>` would be `PhantomData<ForceNoCell<T>>`? I feel like something like `MaskCell` or `HideCell` better expresses what `ForceNoCell` does.

Reviewer 9: It's named by analogy to [`ForceOverwrite<T>`](https://smallcultfollowing.com/babysteps/blog/2024/09/26/overwrite-trait/) and `ForcePin<T>`.

Reviewer 3: Yeah but `NoCell` is a negated term so I find it harder to make sense of with `ForceNoCell`.

Reviewer 9: I had been looking forward to suggesting `ForceFreeze`.

Reviewer 9: More seriously, for `PhantomCell`, it made sense to turn it around, because it means `PhantomNotNoCell`, so we're removing a double negative.

For `ForceX`, though, we mean that it does always implement the `X`.  So probably I'd suggest keeping the orientation in these cases.

Reviewer 3: `ForceOverwrite<T>` as described in that blog post is actually a container holding a `T`, so the analogy anyway doesn't really hold. Your type should be called `PhantomForceNoCell`, if anything.

At that point I think I'd rather associate the naming with the fact that this acts like a pointer indirection.

Reviewer 9: In my view, it's not exactly a `Phantom*` thing, because I see the "phantom" in the name there as related to the fact that you can construct a value.  It's that value that is the "phantom".  Here, you can't construct a `ForceNoCell<T>` value at all (without wrapping it in `PhantomData`, like anything else).

The way I see it is that `ForceUnpin<T>` (and `ForceOverwrite<T>`) are a bit special in that because of the nature of `Unpin` (and `Overwrite`), you actually can construct a value of the trivial wrapper type that always implements the trait.  But that's not true for all auto traits.  What is true for all auto traits is that you can express a type that always implements the trait.

So I see some value in being consistently able to express a type that always implements the auto trait (and passes though other impls) even if you can't always construct a value of that type, as that depends on the nature of the trait whether that's possible for a trivial wrapper.

Reviewer 3: `ForceOverwrite<T>` and your proposed `PhantomNoCell<T>` are completely different types -- one is a `T`, the other is a `!`. One is a container, the other only makes sense inside `PhantomData`. It makes no sense at all to me that they would have a similar name, there's nothing consistent about that.

I'm not opposed to having a family of `Force*`, but they should all be the same.

Reviewer 9: Maybe I wasn't clear?  I'm not drawing a connection here between `ForceOverwrite<T>` and `PhantomNoCell<T>`.

I'm suggesting that, for each auto trait `X`, there should be both a `ForceX<T>` and a `PhantomNotX`.  The former gives people the ability to build up a phantom type that expresses a set of auto traits unconditionally, and the latter gives people the ability to opt-out of a set of auto traits.

Reviewer 9: > `ForceOverwrite<T>` and your proposed `PhantomNoCell<T>` are completely different types...

You mean `ForceOverwrite<T>` and `ForceNoCell<T>` I presume.  I'm going to use `ForceUnpin<T>` below as the comparable, since it's expressible today and also has a reversed polarity like `NoCell`.

> ...are completely different types -- one is a T, the other is a !.

The types that seem relevant here are `ForceUnpin<T>` and `ForceNoCell<T>`.  Neither of those is of type `T` or `!`.

Also, while I expressed the field type for `ForceNoCell<T>` above as `!`, I think it'd be fine for it to be `T`.  And actually, maybe it'd be fine (but probably useless) even for it to have a constructor.  There just can't be an accessor to that field like there can be for `ForceUnpin<T>`.

> One is a container, the other only makes sense inside `PhantomData`. It makes no sense at all to me that they would have a similar name, there's nothing consistent about that.

Even setting aside that they both could be containers (as above), they're both type constructors.  They both can be used within `PhantomData` due to that.  But as it happens, only one of them admits a value constructor with an accessor to that value.

It just so happens that `ForceUnpin` is rather unique in that it soundly admits a value constructor and accessor with a trivial wrapper.  The other auto traits don't.

I suppose if it came down to it, I'd be OK with having a `ForceUnpin` that didn't have a value constructor with an accessor either, if that's what people felt was needed to make this consistent.  We could then add a `ForceUnpinWithAccessor` or whatever, that would be exactly like `ForceUnpin`, but would have the value constructor and accessor.  But that wouldn't make a lot of sense to me.

Reviewer 3: > You mean ForceOverwrite<T> and ForceNoCell<T> I presume. 

Ah yes I do, sorry. :)

> While I expressed the field type for ForceNoCell<T> above as !, I think it'd be fine for it to be T. 

I'd object much less to that from a `Force*` consistency perspective. On the flip side now we get some gnarly aliasing model questions, since so far it is impossible to ever "mask" `Freeze` and the models we have rely on that.

> It seems fine to me for the accessors and other such things to vary between these based on what makes sense given the trait and its contract while keeping the same naming scheme

Agreed.

I still quite like `PhantomDataIndirect` though since it expresses intent directly, rather than indirectly via "I know this is indirect so I know I can soundly hide all the `UnsafeCell` in there".

Most crates that would observe this change could replace their usage of `core::marker::PhantomData<T>` by `core::marker::PhantomData<Ptr<T>>` where `Ptr<T>`
is a pointer-type with the relevant caracteristics regarding lifetime, `Send` and `Sync`ness.

The author doesn't have the necessary knowledge to implement this change, which should still be subject to a crater run to ensure no valid use-cases were missed.

This behaviour change shall be introduced at the same time as the stabilization of `NoCell` in bounds.

## `core::marker::PhantomCell`

> This ZST is proposed as a means for maintainers to reliably opt out of `NoCell` without constraining currently `!NoCell` ZSTs to remain so.

Reviewer 10: Is there any situation where `PhantomCell` is a necessity, other than to prevent future API breakage beforehand? IIUC it's always required to use `UnsafeCell` to create internal mutability, thus it's still unsound to add a `_marker: PhantomCell` and expect to do anything creative with this struct (like mutating other fields behind `&`), yes?

If so, `NoCell`'s behavior sounds like a diverge from soundness guarantees and optimization invariants (`Freeze`). People who write `PhantomCell` would be only to opt-out their structs from const promotions, and they won't expect a surprising de-optimization of all other code using `&Struct`. Also in this way, user has no reason to override `Freeze` impl, making it unnecessary to be exposed to users at all.
